docs/configuration.md
---
title: Configuration
icon: material/cog
---

# Configuration

Configuration of the service is done via environment variables.

## General Configuration

| Name                         | Type                          | Default                        | Description                                                                                    |
| ---------------------------- | ----------------------------- | ------------------------------ | ---------------------------------------------------------------------------------------------- |
| `DOLCE_LOG_LEVEL`            | Deno Log Level [^1]           | `INFO`                         | Loglevel of the service                                                                        |
| `DOCKER_HOST`                | `string?`                     | `/var/run/docker.sock`         | Path to the docker socket or an `ip:port`-pair when used with `DOCKER_TRANSPORT=tcp`           |
| `DOCKER_TRANSPORT`           | `unix` \| `tcp`               | `unix`                         | Transport used to talk to docker                                                               |
| `DOLCE_IDENTIFIER_LABEL`     | `string?`                     | `dolce.identifier`             | See [Container Identifiers](#container-identifiers)                                            |
| `DOLCE_SUPERVISION_MODE`     | SupervisorMode [^2]           | `ALL`                          | See [Supervision Mode](#supervision-mode)                                                      |
| `DOLCE_SUPERVISION_LABEL`    | `string?`                     | `dolce.enabled`                | See [Supervision Mode](#supervision-mode)                                                      |
| `DOLCE_SUPERVISION_PREFIX`   | `string?`                     | `temp-`                        | See [Supervision Mode](#supervision-mode)                                                      |
| `DOLCE_ACTOR_IDENTIFIER`     | `name` \| `image`             | `name`                         | See [Container Identifiers](#container-identifiers)                                            |
| `DOLCE_EVENTS`               | Container Action[] [^3]       | All available                  | See [Event Selection](#event-selection)                                                        |
| `DOLCE_BLACKOUT_WINDOWS`     | [PlainTime, PlainTime][] [^3] | []                             | See [Blackout Times](#blackout-times)                                                          |
| `DOLCE_MIN_TIMEOUT`          | `number`                      | 10                             | See [Notification Backoff](./advanced/notification-backoff.md)                                 |
| `DOLCE_MAX_TIMEOUT`          | `number`                      | 60*60*24                       | See [Notification Backoff](./advanced/notification-backoff.md)                                 |
| `DOLCE_MULTIPLIER`           | `number`                      | 10                             | See [Notification Backoff](./advanced/notification-backoff.md)                                 |
| `DOLCE_RUN_DIRECTORY`        | `string`                      | `/var/run/dolce/`              | Path where the event registry and [lockfile](#lockfile) are stored                             |
| `DOLCE_CUSTOM_TEMPLATE_PATH` | `string?`                     | `/var/dolce-custom-templates/` | See [Custom Templates](./advanced/custom-templates.md)                                         |
| `DOLCE_DEBUG`                | `boolean?`                    | `false`                        | Set to `true` during development to avoid problems with the [Lockfile](#lockfile) in watchmode |

[^1]: [Deno Log Level](https://deno.land/std@0.202.0/log/mod.ts?s=LogLevels)

[^2]: [SupervisionMode](https://github.com/search?q=repo%3Adangrie158/dolce%20SupervisorMode&type=code)

[^3]: [Container Action](https://docs.docker.com/engine/api/v1.27/#tag/System/operation/SystemEvents)

## Supervision Mode

Using the `DOLCE_SUPERVISION_MODE` environment variable you can control which containers are supervised by Dolce. If set
to any other value than `ALL` (the default), Dolce will ignore events by containers that do not match the rules
described below:

### `TAGGED` and `UNTAGGED`

If the variable `DOLCE_SUPERVISION_MODE` is set to the string `TAGGED`, only containers with the tag specified in
`DOLCE_SUPERVISION_LABEL` and a value of **exactly** `true` will create events (Note that docker tag values are always
strings, so no YAML booleans and the (Norway problem)[https://hitchdev.com/strictyaml/why/implicit-typing-removed/]).

`UNTAGGED` is the logical inverse: only containers without the tag set to exactly `true` will create events. In this
case it is useful to change the default value of `DOLCE_SUPERVISION_LABEL` to e.g. `dolce.disabled` to have a meaningful
name for the tag, although this is not required. Go ahead and confuse yourself :wink:

The value of `DOLCE_SUPERVISION_PREFIX` and container names is in both cases irrelevant.

```yaml title="Example using DOLCE_SUPERVISION_MODE=TAGGED"
services:
    dolce:
        image: dangrie158/dolce
            ...
        environment:
            DOLCE_SUPERVISION_MODE: TAGGED
            DOLCE_SUPERVISION_LABEL: monitor.enable
    importantservice:
        labels:
            monitor.enable: true # (1)!
    unimportantservice:
        labels:
            monitor.enable: false # (2)!
    anotherunimportantservice:
        ... # (3)!
```

1. will create notifications, the label is set to exactly "true"
2. won't create notifications, the label is explicitly set to a value other than "true"
3. won't create notifications, the label is not set at all

### `PREFIXED` and `NOTPREFIXED`

If the variable `DOLCE_SUPERVISION_MODE` is set to the string `PREFIXED`, only containers with a name that starts with
the value of `DOLCE_SUPERVISION_PREFIX` will create events. Events by all other containers are ignored.

`NOTPREFIXED` is the logical inverse: all containers will create events **except** they are prefixed with the value in
`DOLCE_SUPERVISION_PREFIX`.

The value of `DOLCE_SUPERVISION_LABEL` and any container tags is in both cases irrelevant.

## Container Identifiers

By default Dolce will use the container name in notifications to identify the container that changed the state. If you
set `DOLCE_ACTOR_IDENTIFIER` to `image`, the name of the dockerimage is used instead. You can also specify a custom
identifier per container by adding a label with the key specified in `DOLCE_IDENTIFIER_LABEL` to your container.

For example:

```yaml title="Example using DOLCE_ACTOR_IDENTIFIER=image and a custom DOLCE_IDENTIFIER_LABEL"
services:
    dolce:
        image: dangrie158/dolce
            ...
        environment:
            DOLCE_ACTOR_IDENTIFIER: image
            DOLCE_IDENTIFIER_LABEL: dolce.label # (1)!
    importantservice:
        labels:
            dolce.label: "my_important_service_do_not_kill" # (2)!
    anotherimportantservice:
        ... # (3)!
```

1. `dolce.identifier` by default
2. using the label specified in `DOLCE_IDENTIFIER_LABEL`, this container will be referred to as
   `my_important_service_do_not_kill` in any notifications
3. This container will be referred to by its image name due to the setting of `DOLCE_ACTOR_IDENTIFIER=image`

## Event Selection

By default you will get notified for a subset of all available events that are fired by the Docker Engine: `start`

- `die`
- `kill`
- `oom`
- `stop`
- `pause`
- `unpause`
- `health_status`

You can narrow this list down or expand it by setting the `DOLCE_EVENTS` setting. If you are for example only interested
in `die` and `start` events you can set:

```yaml
dolce:
    image: dangrie158/dolce
        ...
    environment:
        DOLCE_EVENTS: die,start
```

This is the list of all available events:

`attach`, `commit`, `copy`, `create`, `destroy`, `detach`, `die`, `exec_create`, `exec_detach`, `exec_start`, `export`,
`health_status`, `kill`, `oom`, `pause`, `rename`, `resize`, `restart`, `start`, `stop`, `top`, `unpause`, `update`.

## Blackout Times

You can specify a list of time ranges in which all events should be ignored. This is useful if you have a maintenance
window or if you regularly stop and start containers for some reason. The format is `HH:[MM[:SS]]-HH:[MM[:SS]]` and you
can specify multiple ranges by separating them with a comma.

```yaml
dolce:
    image: dangrie158/dolce
        ...
    environment:
        DOLCE_BLACKOUT_WINDOWS: 02:00-04:00,22:00-23:59
```

If the blackout window ends, the status of all containers is compared to the last known state and notifications are sent
if any container is in a different state than before. This enables you to schedule automatic container updates in a
specified window and still get notified if something goes wrong during the update process.

## Run Directory

If you plan on using multiple instances of Dolce (e.g. to enable multiple configurations) make sure that if this is
mounted in a volume, no two different services access the same file.

### Lockfile

The lockfile is used to detect unexpected shutdowns and avoid mutliple instances of the service running at the same
time. It is located at `${DOLCE_RUN_DIRECTORY}/lockfile` You can mount this path in a volume to detect unexpected
shutdowns outside of the lifetime of the container.

## Use Local Timezone for Eventdates

By default, dolce uses the UTC timezone for all eventdates. This is a good default because it is the same timezone that
the docker API uses. However, if you want to use the local timezone of the host, you can set the `TZ` environment
variable to the desired timezone.

```yaml
services:
  dolce:
    image: dangrie158/dolce:v2.10.7
    restart: unless-stopped
    environment:
      ...
      TZ: Europe/Berlin
```


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

#### [v3.4.0](https://github.com/dangrie158/dolce/compare/v3.3.0...v3.4.0)

> 20 March 2025

#### [v3.3.0](https://github.com/dangrie158/dolce/compare/v3.2.0...v3.3.0)

> 23 November 2024

- feat: Switch to alpine distro [`#28`](https://github.com/dangrie158/dolce/pull/28)
- Add support for STARTTLS [`#26`](https://github.com/dangrie158/dolce/pull/26)
- docs: explain usage of `SMTP_USETLS` to use `STARTTLS`
  [`11561b4`](https://github.com/dangrie158/dolce/commit/11561b4fed5d0dc310b1074546fbf391b0553a50)
- Switch to alpine distro
  [`7b5870a`](https://github.com/dangrie158/dolce/commit/7b5870a678a38f518b18ac5f5f499dcd69dd510c)
- Apply method naming convention
  [`68c7450`](https://github.com/dangrie158/dolce/commit/68c7450c178d5fbee1209bcd3d10ece08b95360b)

#### [v3.2.0](https://github.com/dangrie158/dolce/compare/v3.1.0...v3.2.0)

> 5 October 2024

- feat: notifiy about state changes during blackout windows
  [`5832cfe`](https://github.com/dangrie158/dolce/commit/5832cfe9ac492dec731366bd241258c66c547a04)
- docs: better documentation for blackout window feature
  [`5cdfbba`](https://github.com/dangrie158/dolce/commit/5cdfbba7330a20e49b61a13c5e5bc01f3bbde2ed)
- feat: add docs task to deno.jsonc
  [`2ccaac2`](https://github.com/dangrie158/dolce/commit/2ccaac2e125eb63ae59ec7d378f0ed4928575b44)

#### [v3.1.0](https://github.com/dangrie158/dolce/compare/v3.0.2...v3.1.0)

> 4 October 2024

#### [v3.0.2](https://github.com/dangrie158/dolce/compare/v3.0.1...v3.0.2)

> 4 October 2024

- fix: make SMTP_USER and SMTP_PASSWORD optional again
  [`b4b45f0`](https://github.com/dangrie158/dolce/commit/b4b45f088c6ab56d153dd31feb3829cc68a48edd)

#### [v3.0.1](https://github.com/dangrie158/dolce/compare/v3.0.0...v3.0.1)

> 20 September 2024

- fix: use correct event timestamp in ms instead of seconds
  [`08d80ce`](https://github.com/dangrie158/dolce/commit/08d80ced3b28b4b42ced1ee1963b01a0a66161dc)
- docs: fix link to Blackout Times configuration section
  [`6f14894`](https://github.com/dangrie158/dolce/commit/6f1489405ff43525b3896982f9af1569103319fa)

### [v3.0.0](https://github.com/dangrie158/dolce/compare/v2.10.9...v3.0.0)

> 20 September 2024

- feat: implement DOLCE_BLACKOUT_WINDOWS function
  [`763dd33`](https://github.com/dangrie158/dolce/commit/763dd334a1f446e4cb8d6e07374a1f40571aa45f)
- feat: add transformers option to ConfigOption interface
  [`fddbc4c`](https://github.com/dangrie158/dolce/commit/fddbc4c7d4bf778964a8c65dd95884f128d71cb5)
- feat: introduce new DOLCE_RUN_DIRECTORY to make deno event registry path configurable
  [`0f68ce5`](https://github.com/dangrie158/dolce/commit/0f68ce5cbb1f8e846496b2d3658b8a19d82c7e97)

#### [v2.10.9](https://github.com/dangrie158/dolce/compare/v2.10.8...v2.10.9)

> 19 July 2024

- docs: add section for using local datetimes for event dates
  [`8f3693b`](https://github.com/dangrie158/dolce/commit/8f3693b2aeb417a8678da72fb11460632c9a0b2c)
- docs: remove deprecated version attribute from example configs
  [`d87023a`](https://github.com/dangrie158/dolce/commit/d87023a954b1b21281cd70c5b5cb29a1d15b9c00)

#### [v2.10.8](https://github.com/dangrie158/dolce/compare/v2.10.7...v2.10.8)

> 19 July 2024

- fix: use inline styles for mail template to improve mail client compatability
  [`1257a61`](https://github.com/dangrie158/dolce/commit/1257a61b281938d37a4e8f4a988c81a3d6e409eb)
- fix: improve typing for HttpSocket.read_response()
  [`86485c0`](https://github.com/dangrie158/dolce/commit/86485c0afcf8351a32411f6fcb616a13a70e01a3)

#### [v2.10.7](https://github.com/dangrie158/dolce/compare/v2.10.6...v2.10.7)

> 19 April 2024

#### [v2.10.6](https://github.com/dangrie158/dolce/compare/v2.10.5...v2.10.6)

> 19 April 2024

#### [v2.10.5](https://github.com/dangrie158/dolce/compare/v2.10.4...v2.10.5)

> 19 April 2024

#### [v2.10.4](https://github.com/dangrie158/dolce/compare/v2.10.3...v2.10.4)

> 19 April 2024

#### [v2.10.3](https://github.com/dangrie158/dolce/compare/v2.10.2...v2.10.3)

> 19 April 2024

#### [v2.10.2](https://github.com/dangrie158/dolce/compare/v2.10.1...v2.10.2)

> 19 April 2024

- fix: error when automatically generating release notes
  [`a4debc3`](https://github.com/dangrie158/dolce/commit/a4debc318bb570260b96e9f2bad3b46fc64293a3)

#### [v2.10.1](https://github.com/dangrie158/dolce/compare/v2.10.0...v2.10.1)

> 19 April 2024

- fix: don't throw an error when explicitly setting a DOCKER_HOST with DOCKER_TRANSPORT=unix
  [`19a83be`](https://github.com/dangrie158/dolce/commit/19a83bead1bccdb37cb533ed11417536f0b21d5a)

#### [v2.10.0](https://github.com/dangrie158/dolce/compare/v2.9.0...v2.10.0)

> 2 April 2024

- feat: use official denoland images which now support aarch64
  [`bf53424`](https://github.com/dangrie158/dolce/commit/bf5342491533826dd7d9c21ef3d34e53264f4701)
- fix: deprecation warning for not using fine-grained unstable flags
  [`3bde9fa`](https://github.com/dangrie158/dolce/commit/3bde9fa862458c74c43ff4a862778565f1ffb329)

#### [v2.9.0](https://github.com/dangrie158/dolce/compare/v2.8.0...v2.9.0)

> 24 March 2024

- fix: try to reconnect if the eventstream got unexpectedly closed
  [`ee018a3`](https://github.com/dangrie158/dolce/commit/ee018a3d268e5a91841984f1f97e735441a26d08)
- fix: off by one error that causes events to get delivered multiple times
  [`3224e08`](https://github.com/dangrie158/dolce/commit/3224e08cceda26790a0159141f48ed8b94e3a53b)
- fix: print active configuratoin on startup instead of defaults
  [`56e0a01`](https://github.com/dangrie158/dolce/commit/56e0a01b9a84f640a3c5fe2191ec1414be4afb6e)

#### [v2.8.0](https://github.com/dangrie158/dolce/compare/v2.7.0...v2.8.0)

> 29 October 2023

#### [v2.7.0](https://github.com/dangrie158/dolce/compare/v2.6.2...v2.7.0)

> 17 October 2023

- feat: add DOLCE_SUPERVISION_MODE s PREFIXED, NOTPREFIXED and UNTAGGED

configuration.ts
import * as log from "@std/log";
import { CheckedConfiguration, ConfigOption } from "./lib/env.ts";
import { CONTAINER_ACTIONS, ContainerAction } from "./lib/docker-api.ts";

const SUPERVISION_MODES = ["TAGGED", "UNTAGGED", "PREFIXED", "NOTPREFIXED", "ALL"] as const;
type SupervisorMode = (typeof SUPERVISION_MODES)[number];
type ActorIdentifier = "image" | "name";

function timewindow_converter(value: string): [Temporal.PlainTime, Temporal.PlainTime] {
    const [start, end] = value.split("-").map((time) => Temporal.PlainTime.from(time));
    if (start === undefined || end === undefined) {
        throw new Error(`Invalid time window format: ${value}, should be HH[:MM[:SS]]-HH:[:MM[:SS]]`);
    }
    return [start, end];
}

export class Configuration extends CheckedConfiguration {
    @ConfigOption({ env_variable: "DOLCE_LOG_LEVEL", one_of: log.LogLevelNames })
    static readonly loglevel: log.LevelName = "INFO";

    @ConfigOption()
    static readonly docker_host?: string = "/var/run/docker.sock";

    @ConfigOption({ one_of: ["unix", "tcp"] })
    static readonly docker_transport: string = "unix";

    @ConfigOption({ env_variable: "DOLCE_IDENTIFIER_LABEL" })
    static readonly identifier_label: string = "dolce.identifier";

    @ConfigOption({
        env_variable: "DOLCE_SUPERVISION_MODE",
        one_of: SUPERVISION_MODES,
    })
    static readonly supervision_mode: SupervisorMode = "ALL";

    @ConfigOption({ env_variable: "DOLCE_SUPERVISION_LABEL" })
    static readonly supervision_label: string = "dolce.enabled";

    @ConfigOption({ env_variable: "DOLCE_SUPERVISION_PREFIX" })
    static readonly supervision_prefix: string = "temp-";

    @ConfigOption({ env_variable: "DOLCE_ACTOR_IDENTIFIER", one_of: ["image", "name"] })
    static readonly actor_identifier: ActorIdentifier = "name";

    @ConfigOption({ type: Array, env_variable: "DOLCE_EVENTS", some_of: CONTAINER_ACTIONS })
    static readonly events: ContainerAction[] = [
        "start",
        "die",
        "kill",
        "oom",
        "stop",
        "pause",
        "unpause",
        "health_status",
    ];

    @ConfigOption({ type: Number, env_variable: "DOLCE_MIN_TIMEOUT" })
    static readonly min_timeout: number = 10;

    @ConfigOption({ type: Number, env_variable: "DOLCE_MAX_TIMEOUT" })
    static readonly max_timeout: number = 60 * 60 * 24;

    @ConfigOption({ type: Number, env_variable: "DOLCE_MULTIPLIER" })
    static readonly multiplier: number = 10;

    @ConfigOption({ env_variable: "DOLCE_RUN_DIRECTORY" })
    static readonly run_directory: string = "/var/run/dolce/";

    @ConfigOption({ env_variable: "DOLCE_CUSTOM_TEMPLATE_PATH" })
    static readonly custom_template_path: string = "/var/dolce-custom-templates/";

    @ConfigOption({
        type: Array,
        env_variable: "DOLCE_BLACKOUT_WINDOWS",
        transformer: (values: string[]) => values.map((value: string) => timewindow_converter(value)),
    })
    static readonly blackout_windows: [Temporal.PlainTime, Temporal.PlainTime][] = [];

    @ConfigOption({ type: Boolean, env_variable: "DOLCE_DEBUG" })
    static readonly debug: boolean = false;

    override toString() {
        return `\nConfiguration {
            ${
            Object.entries(Configuration)
                .map(([key, value]) => `\t${key}: ${value}`)
                .join(",\n")
        }
        }`;
    }
}

log.setup({
    handlers: {
        default: new log.ConsoleHandler(Configuration.loglevel, {
            formatter: (record) => `${record.levelName}\t${record.loggerName}\t ${record.msg}`,
        }),
    },
    loggers: {
        main: { level: Configuration.loglevel, handlers: ["default"] },
        notifier: { level: Configuration.loglevel, handlers: ["default"] },
        lockfile: { level: Configuration.loglevel, handlers: ["default"] },
    },
});
